Example 3d cylinder (#1570)
chaos-polymtl/lethe: doc/source/examples/incompressible-flow/3d-turbulent-flow-around-cylinder/3d-turbulent-flow-around-cylinder.rst

.. code-block:: text
  
  subsection mesh
    set type               = dealii
    set grid type          = custom_channel_with_cylinder
    set grid arguments     = 25 : 8 : 52 : 4.71238898038 : 4 : 0.75 : 5 : 1:  false: true
    set initial boundary refinement = 1 
    set boundaries refined = 2
  end

.. note::

  This grid generator is only present in the 9.7 version of the deal.II library.

Box refinement
~~~~~~~~~~~~~~~~
The ``box refinement`` subsection refines the mesh in a specific region of the domain. In this case, we refine the mesh around the cylinder to capture the boundary layer and wake region more accurately. The box refinement is defined by its center, size, and refinement level. The mesh is refined by the box refinement twice by setting ``ìnitial refinement = 2``.

.. code-block:: text

  subsection box refinement
    subsection mesh
      set type               = dealii
      set grid type          = subdivided_hyper_rectangle
      set grid arguments     = 1,1,1 : -2, -3, -1 : 52,3,5 : false
      set initial refinement = 0
    end
    set initial refinement = 2

>   end

Author: Sure. Should the final mesh picture be in the results section? Amongst the parameter it might be a bit intrusive.

.. note::

  THe simulation takes approximatively 10 hours on 16 cores of a AMD Ryzen 9 7950X 16-Core Processor.

----------------------
Results and Discussion
----------------------

In the following, results obtained with a box refinement of [2,3,4] and using :math:`Q_1Q_1` and :math:`Q_2Q_2` elements are presented. The results are compared with the literature, including the work of Cardell [#Cardell1993]_, Ong and Wallace [#Ong1996]_, and Norberg [#Norberg1987]_.

First, the following animation displays the evolution of the velocity magnitude on a slice of the domain over time for a very coarse mesh (:math:`Q_1Q_1` with box refinement of 2):

+----------------------------------------------------------------------------------------------------------------------------------------------------+
| .. raw:: html                                                                                                                                      |
|                                                                                                                                                    |

> |    <iframe width="700" height="400" src="https://www.youtube.com/embed/ojhKCPY8Bho?si=wmkUBWQyEQIC3cn"  frameborder="0" allowfullscreen></iframe>  |

Author: you mean increase the size of the video or make a new video which is zoomed in on the cylinder?

> |    <iframe width="700" height="400" src="https://www.youtube.com/embed/ojhKCPY8Bho?si=wmkUBWQyEQIC3cn"  frameborder="0" allowfullscreen></iframe>  |

Reviewer: Well the margins are really large and the cylinder is small, so I would zoom on the cylinder a bit so we see better the detail of the wake also.

> |    <iframe width="700" height="400" src="https://www.youtube.com/embed/ojhKCPY8Bho?si=wmkUBWQyEQIC3cn"  frameborder="0" allowfullscreen></iframe>  |

Author: Ok yeah. I might have to rerun the simulation for this, so I might keep it for a next PR if you don't mind.

  :class: copy-button

  python postprocess-cylinder.py -f output -l label

where the ``-f`` option specifies the output folder and the ``-l`` option specifies the label of the simulation (e.g., ``Q1Q1`` or ``Q2Q2``). This script can be used to visualize a single simulation or to compare multiple simulations by providing a list of folders and labels.

The drag coefficient measured experimentally by Norberg [#Norberg1994]_ is 1.0075. The Strouhal number value reported by Cardell [#Cardell1993]_ is :math:`0.215 \pm 0.005` and the value reported by Ong and Wallace [#Ong1996]_ is :math:`0.208 \pm 0.002`.

The following table summarizes the results obtained in this example, including the Strouhal number :math:`S_t` and the drag coefficient :math:`C_d`. 

.. list-table::
   :widths: 10 10 10 10
   :header-rows: 1

   * - Element
     - Mesh refinement (l)
     - :math:`C_d`
     - :math:`S_t`
   * - :math:`Q_1Q_1`
     - 2
     - 1.2105
     - 0.1965
   * - :math:`Q_1Q_1`
     - 3
     - 1.0305
     - 0.2110
   * - :math:`Q_1Q_1`
     - 4 
     - 0.9901
     - 0.2122
   * - :math:`Q_2Q_2`
     - 2
     - 1.0008
     - 0.2134
   * - :math:`Q_2Q_2`
     - 3
     - 0.9784

>      - 0.2112

Author: Nah it's just too long to run. l4 is around 10M cells so I did not run it. I will run it soon though, will make a subsenquent Pr for that :)

--------------
Parameter File
--------------

Mesh
~~~~

The ``mesh`` subsection specifies the computational grid. We use a custom mesh generated using the functionality `custom_channel_with_cylinder`  of the deal.II library's `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ to discretize the domain using high-order elements. The mesh is also refined once in the vicinity of the cylinder to capture the boundary layer and wake region more accurately. This is achieved by setting ``initial boundary refinement = 1`` and ``boundaries refined = 2``. 

.. code-block:: text
  
  subsection mesh
    set type                        = dealii
    set grid type                   = uniform_channel_with_cylinder
    set grid arguments              = 25 : 8 : 52 : 4.71238898038 : 4 : 0.75 : 5 : 1:  false: true
    set initial boundary refinement = 1 
    set boundaries refined          = 2
  end

> .. note::

Author: ```suggestion
.. warning::
```

  :name: geometry
  :height: 9cm

> Without loss of generality, :math:`u_\infty` is set to 1 and the cylinder diameter :math:`D` is set to 1. The Reynolds number is defined as :math:`Re = \frac{U_{\infty}D}{\nu}`, where :math:`\nu` is the kinematic viscosity. The kinematic viscosity is set to :math:`\nu=2.5641025e-04`, which results in a Reynolds number of 3900.

Reviewer: ```suggestion
Without loss of generality, :math:`U_\infty` is set to 1 and the cylinder diameter :math:`D` is set to 1. The Reynolds number is defined as :math:`Re = \frac{U_{\infty}D}{\nu}`, where :math:`\nu` is the kinematic viscosity. The kinematic viscosity is set to :math:`\nu=2.5641025\times 10^{-4}`, which results in a Reynolds number of 3900.
```

--------------
Parameter File
--------------

Mesh
~~~~

> The ``mesh`` subsection specifies the computational grid. We use a custom mesh generated using the functionality `custom_channel_with_cylinder`  of the deal.II library's `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ to discretize the domain using high-order elements. The mesh is also refined once in the vicinity of the cylinder to capture the boundary layer and wake region more accurately. This is achieved by setting ``initial boundary refinement = 1`` and ``boundaries refined = 2``. 

Reviewer: ```suggestion
The ``mesh`` subsection specifies the computational grid. We use a custom mesh generated using the functionality ``custom_channel_with_cylinder``  of the deal.II library's `GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html>`_ to discretize the domain using high-order elements. The mesh is also refined once in the vicinity of the cylinder to capture the boundary layer and wake region more accurately. This is achieved by setting ``initial boundary refinement = 1`` and ``boundaries refined = 2``. 
```

.. code-block:: text
  
  subsection mesh
    set type                        = dealii
    set grid type                   = uniform_channel_with_cylinder

>     set grid arguments              = 25 : 8 : 52 : 4.71238898038 : 4 : 0.75 : 5 : 1:  false: true

Reviewer: ```suggestion
    set grid arguments              = 25 : 8 : 52 : 4.71238898038 : 4 : 0.75 : 5 : 1:  false : true
```

    set initial boundary refinement = 1 
    set boundaries refined          = 2
  end

.. note::

>   This `uniform_channel_with_cylinder` grid generator is only present in the 9.7 version of the deal.II library.

Reviewer: ```suggestion
  This ``uniform_channel_with_cylinder`` grid generator is only present in the 9.7 version of the deal.II library.
```

|                                                                                                                                                    |
+----------------------------------------------------------------------------------------------------------------------------------------------------+

The key validation metrics are:

- **Strouhal number**:

.. math::

  St = \frac{f D}{U_\infty}

> where :math:`f` is the frequency of vortex shedding, :math:`D` is the cylinder diameter, and :math:`U_\infty` is the free-stream velocity. T The frequency of vortex shedding is determined using a Fast Fourier Transform (FFT) of the lift force.

Reviewer: ```suggestion
where :math:`f` is the frequency of vortex shedding, :math:`D` is the cylinder diameter, and :math:`U_\infty` is the free-stream velocity. The frequency of vortex shedding is determined using a Fast Fourier Transform (FFT) of the lift force.
```
